File: src/remedy/connect/relay.py
"""Owner-run Grove Connect relay.

TCP listen on a **chosen IPv4**. Clients send a 16-byte session id, then framed
blobs (``u32be length | rest``). Two peers that share a session id are spliced:
bytes are copied, never decrypted, never logged.

Wildcard binds (``0.0.0.0`` / ``*`` / ``::``) are refused.
"""

from __future__ import annotations

import argparse
import contextlib
import logging
import socket
import struct
import sys
import threading
import time
from collections.abc import Callable

from remedy.connect.bind import assert_chosen_bind as _assert_bind
from remedy.connect.bind import is_chosen_ipv4

log = logging.getLogger(__name__)

SESSION_ID_LEN = 16
MAX_PAYLOAD = 65536  # 64 KiB
MAX_FRAME = MAX_PAYLOAD + 4
DEFAULT_HOST = "127.0.0.1"
DEFAULT_PORT = 7402
# How long a lone peer waits for its counterpart. Tests keep this short;
# production host reconnects, so a few minutes covers a phone opening the app.
WAIT_PEER_S = 60.0


def assert_chosen_bind(host: str) -> str:
    """Refuse wildcard / non-IPv4 binds. Relays listen on a chosen IPv4 only."""
    text = _assert_bind(host)
    if not is_chosen_ipv4(text):
        raise ValueError(f"connect-relay needs a chosen IPv4, not {host!r}")
    return text


def _recvall(sock: socket.socket, n: int) -> bytes | None:
    buf = bytearray()
    while len(buf) < n:
        try:
            chunk = sock.recv(n - len(buf))
        except OSError:
            return None
        if not chunk:
            return bytes(buf) if buf else None
        buf.extend(chunk)
    return bytes(buf)


class _Slot:
    __slots__ = ("lock", "peers")

    def __init__(self) -> None:
        self.lock = threading.Lock()
        self.peers: list[socket.socket] = []


def _copy_frames(src: socket.socket, dst: socket.socket) -> None:
    """Copy framed blobs src → dst. Does not inspect or log ``rest``."""
    try:
        while True:
            header = _recvall(src, 4)
            if header is None or len(header) < 4:
                break
            length = struct.unpack("!I", header)[0]
            if length > MAX_PAYLOAD:
                break
            payload = b"" if length == 0 else _recvall(src, length)
            if payload is None or len(payload) != length:
                break
            try:
                dst.sendall(header + payload)
            except OSError:
                break
    except OSError:
        pass
    finally:
        with contextlib.suppress(OSError):
            dst.shutdown(socket.SHUT_WR)


def _handle_client(
    conn: socket.socket,
    sessions: dict[bytes, _Slot],
    sessions_lock: threading.Lock,
    wait_peer_s: float = 60.0,
) -> None:
    conn.settimeout(60.0)
    sid = _recvall(conn, SESSION_ID_LEN)
    if sid is None or len(sid) != SESSION_ID_LEN:
        conn.close()
        return
    with sessions_lock:
        slot = sessions.get(sid)
        if slot is None:
            slot = _Slot()
            sessions[sid] = slot
    peer: socket.socket | None = None
    with slot.lock:
        if len(slot.peers) >= 2:
            conn.close()
            return
        slot.peers.append(conn)
        if len(slot.peers) == 2:
            peer = slot.peers[0] if slot.peers[1] is conn else slot.peers[1]
    if peer is None:
        deadline = time.monotonic() + float(wait_peer_s)
        while time.monotonic() < deadline:
            with slot.lock:
                if len(slot.peers) == 2:
                    peer = slot.peers[0] if slot.peers[1] is conn else slot.peers[1]
                    break
            time.sleep(0.02)
        if peer is None:
            with slot.lock:
                if conn in slot.peers:
                    slot.peers.remove(conn)
            with sessions_lock:
                if sessions.get(sid) is slot and not slot.peers:
                    sessions.pop(sid, None)
            conn.close()
            return
    conn.settimeout(None)
    with contextlib.suppress(OSError):
        peer.settimeout(None)
    try:
        _copy_frames(conn, peer)
    finally:
        with contextlib.suppress(OSError):
            conn.close()
        with slot.lock:
            if conn in slot.peers:
                slot.peers.remove(conn)
        with sessions_lock:
            if sessions.get(sid) is slot and not slot.peers:
                sessions.pop(sid, None)


def _accept_loop(
    listen: socket.socket, stop: threading.Event, wait_peer_s: float
) -> None:
    sessions: dict[bytes, _Slot] = {}
    sessions_lock = threading.Lock()
    while not stop.is_set():
        try:
            conn, _addr = listen.accept()
        except TimeoutError:
            continue
        except OSError:
            if stop.is_set():
                break
            continue
        threading.Thread(
            target=_handle_client,
            args=(conn, sessions, sessions_lock, wait_peer_s),
            daemon=True,
            name="connect-relay-peer",
        ).start()


def start_relay(
    host: str, port: int = 0, *, wait_peer_s: float | None = None
) -> tuple[int, Callable[[], None]]:
    """Bind a chosen IPv4 and serve in a daemon thread. Returns ``(port, stop)``."""
    wait_s = float(WAIT_PEER_S if wait_peer_s is None else wait_peer_s)
    chosen = assert_chosen_bind(host)
    listen = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    listen.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    try:
        listen.bind((chosen, int(port)))
        listen.listen(32)
    except OSError:
        listen.close()
        raise
    listen.settimeout(0.2)
    bound = int(listen.getsockname()[1])
    stop = threading.Event()
    thread = threading.Thread(
        target=_accept_loop,
        args=(listen, stop, wait_s),
        daemon=True,
        name="connect-relay",
    )
    thread.start()
    log.info("connect-relay listening on %s:%s", chosen, bound)

    def _stop() -> None:
        stop.set()
        with contextlib.suppress(OSError):
            listen.close()
        thread.join(timeout=2.0)

    return bound, _stop


def serve(host: str = DEFAULT_HOST, port: int = DEFAULT_PORT) -> None:
    """Blocking listen until KeyboardInterrupt."""
    _bound, stop = start_relay(host, port)
    try:
        while True:
            time.sleep(0.5)
    except KeyboardInterrupt:
        pass
    finally:
        stop()


def main(
    argv: list[str] | None = None,
    *,
    host: str | None = None,
    port: int | None = None,
) -> int:
    """CLI / ``python -m remedy.connect.relay`` entry. Returns a process exit code."""
    if host is None or port is None:
        parser = argparse.ArgumentParser(
            prog="remedy connect-relay",
            description=(
                "Owner-run Grove Connect relay. Forwards framed blobs between two "
                "peers that share a session id. Does not decrypt. Chosen IPv4 only."
            ),
        )
        parser.add_argument("--host", default=DEFAULT_HOST, help="Chosen IPv4 (not 0.0.0.0)")
        parser.add_argument("--port", type=int, default=DEFAULT_PORT)
        ns = parser.parse_args(argv)
        if host is None:
            host = ns.host
        if port is None:
            port = ns.port
    assert host is not None and port is not None
    try:
        bound, stop = start_relay(host, port, wait_peer_s=900.0)
    except ValueError as exc:
        print(f"connect-relay: {exc}", file=sys.stderr)
        return 2
    print(
        f"connect-relay listening on {host}:{bound} "
        "(forwards framed blobs; does not decrypt)",
        file=sys.stderr,
    )
    try:
        while True:
            time.sleep(0.5)
    except KeyboardInterrupt:
        return 0
    finally:
        stop()


if __name__ == "__main__":
    raise SystemExit(main())

File: native/go/connect/relay.go
package connect

import (
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net"
	"strings"
	"sync"
	"time"
)

const (
	// RelayMaxPayload is the max framed blob body (64 KiB). Matches Python MAX_PAYLOAD.
	RelayMaxPayload = 65536
	// RelayDefaultHost is the loopback bind used by the CLI when --host is omitted.
	RelayDefaultHost = "127.0.0.1"
	// RelayWaitPeer is how long a lone peer waits for its counterpart.
	RelayWaitPeer = 60 * time.Second
	// relayAcceptTimeout lets Stop interrupt Accept without spinning forever.
	relayAcceptTimeout = 200 * time.Millisecond
	// relayHelloTimeout bounds the initial 16-byte session-id read.
	relayHelloTimeout = 60 * time.Second
)

// ErrRelay is a Connect relay failure (bind policy, session id, or dial).
var ErrRelay = errors.New("connect relay")

// AssertRelayBind refuses wildcards and non-IPv4 hosts. Relays listen on a chosen IPv4 only.
func AssertRelayBind(host string) (string, error) {
	text, err := AssertChosenBind(host)
	if err != nil {
		return "", err
	}
	if !IsChosenIPv4(text) {
		return "", fmt.Errorf("%w: needs a chosen IPv4, not %q", ErrRelay, host)
	}
	return text, nil
}

type relaySlot struct {
	mu    sync.Mutex
	peers []net.Conn
	wake  chan struct{}
}

// Relay is an owner-run Grove Connect splice: two peers sharing a 16-byte
// session id exchange framed blobs (u32be length | rest). Bytes are copied,
// never decrypted, never logged.
type Relay struct {
	ln       net.Listener
	bindHost string
	bindPort int
	waitPeer time.Duration

	mu       sync.Mutex
	sessions map[string]*relaySlot
	stopCh   chan struct{}
	wg       sync.WaitGroup
}

// StartRelay binds host:port (port 0 = ephemeral) and serves in the background.
// waitPeer <= 0 uses RelayWaitPeer.
func StartRelay(host string, port int, waitPeer time.Duration) (*Relay, error) {
	chosen, err := AssertRelayBind(host)
	if err != nil {
		return nil, err
	}
	if port < 0 || port > 65535 {
		return nil, fmt.Errorf("%w: invalid port %d", ErrRelay, port)
	}
	if waitPeer <= 0 {
		waitPeer = RelayWaitPeer
	}

	addr := net.JoinHostPort(chosen, fmt.Sprintf("%d", port))
	ln, err := net.Listen("tcp", addr)
	if err != nil {
		return nil, err
	}
	tcpAddr, ok := ln.Addr().(*net.TCPAddr)
	if !ok || tcpAddr == nil {
		_ = ln.Close()
		return nil, fmt.Errorf("%w: listener has no TCP address", ErrRelay)
	}

	r := &Relay{
		ln:       ln,
		bindHost: tcpAddr.IP.String(),
		bindPort: tcpAddr.Port,
		waitPeer: waitPeer,
		sessions: make(map[string]*relaySlot),
		stopCh:   make(chan struct{}),
	}
	r.wg.Add(1)
	go r.acceptLoop()
	return r, nil
}

// Port returns the bound TCP port.
func (r *Relay) Port() int {
	return r.bindPort
}

// Addr returns host:port.
func (r *Relay) Addr() string {
	return net.JoinHostPort(r.bindHost, fmt.Sprintf("%d", r.bindPort))
}

// Stop closes the listener and waits for accept/handlers to finish. Idempotent.
func (r *Relay) Stop() {
	if r == nil {
		return
	}
	r.mu.Lock()
	select {
	case <-r.stopCh:
		r.mu.Unlock()
		return
	default:
		close(r.stopCh)
	}
	ln := r.ln
	r.mu.Unlock()
	if ln != nil {
		_ = ln.Close()
	}
	r.wg.Wait()
}

func (r *Relay) acceptLoop() {
	defer r.wg.Done()
	for {
		select {
		case <-r.stopCh:
			return
		default:
		}
		if tcp, ok := r.ln.(*net.TCPListener); ok {
			_ = tcp.SetDeadline(time.Now().Add(relayAcceptTimeout))
		}
		conn, err := r.ln.Accept()
		if err != nil {
			select {
			case <-r.stopCh:
				return
			default:
			}
			if ne, ok := err.(net.Error); ok && ne.Timeout() {
				continue
			}
			continue
		}
		r.wg.Add(1)
		go r.handleClient(conn)
	}
}

func (r *Relay) handleClient(conn net.Conn) {
	defer r.wg.Done()
	defer func() { _ = conn.Close() }()

	_ = conn.SetDeadline(time.Now().Add(relayHelloTimeout))
	sid := make([]byte, SessionIDLen)
	if _, err := io.ReadFull(conn, sid); err != nil {
		return
	}
	_ = conn.SetDeadline(time.Time{})

	key := string(sid)
	slot := r.slotFor(key)

	slot.mu.Lock()
	if len(slot.peers) >= 2 {
		slot.mu.Unlock()
		return
	}
	slot.peers = append(slot.peers, conn)
	var peer net.Conn
	if len(slot.peers) == 2 {
		peer = otherPeer(slot.peers, conn)
		select {
		case <-slot.wake:
		default:
			close(slot.wake)
		}
	}
	slot.mu.Unlock()

	if peer == nil {
		timer := time.NewTimer(r.waitPeer)
		defer timer.Stop()
		select {
		case <-slot.wake:
		case <-timer.C:
			r.dropPeer(key, slot, conn)
			return
		case <-r.stopCh:
			r.dropPeer(key, slot, conn)
			return
		}
		slot.mu.Lock()
		peer = otherPeer(slot.peers, conn)
		slot.mu.Unlock()
		if peer == nil {
			r.dropPeer(key, slot, conn)
			return
		}
	}

	copyFrames(conn, peer)
	r.dropPeer(key, slot, conn)
}

func (r *Relay) slotFor(key string) *relaySlot {
	r.mu.Lock()
	defer r.mu.Unlock()
	slot := r.sessions[key]
	if slot == nil {
		slot = &relaySlot{wake: make(chan struct{})}
		r.sessions[key] = slot
	}
	return slot
}

func (r *Relay) dropPeer(key string, slot *relaySlot, conn net.Conn) {
	slot.mu.Lock()
	for i, p := range slot.peers {
		if p == conn {
			slot.peers = append(slot.peers[:i], slot.peers[i+1:]...)
			break
		}
	}
	empty := len(slot.peers) == 0
	slot.mu.Unlock()
	if empty {
		r.mu.Lock()
		if r.sessions[key] == slot {
			delete(r.sessions, key)
		}
		r.mu.Unlock()
	}
}

func otherPeer(peers []net.Conn, self net.Conn) net.Conn {
	for _, p := range peers {
		if p != self {
			return p
		}
	}
	return nil
}

// copyFrames copies framed blobs src → dst. Does not inspect or log rest.
func copyFrames(src, dst net.Conn) {
	defer halfCloseWrite(dst)

	header := make([]byte, 4)
	for {
		if _, err := io.ReadFull(src, header); err != nil {
			break
		}
		length := binary.BigEndian.Uint32(header)
		if length > RelayMaxPayload {
			break
		}
		var payload []byte
		if length > 0 {
			payload = make([]byte, length)
			if _, err := io.ReadFull(src, payload); err != nil {
				break
			}
		}
		if _, err := dst.Write(header); err != nil {
			break
		}
		if length > 0 {
			if _, err := dst.Write(payload); err != nil {
				break
			}
		}
	}
}

func halfCloseWrite(c net.Conn) {
	type closeWriter interface{ CloseWrite() error }
	if cw, ok := c.(closeWriter); ok {
		_ = cw.CloseWrite()
	}
}

// DialRelay connects to an owner-run relay URL and writes the 16-byte session id.
func DialRelay(ctx context.Context, rawURL string, sessionID []byte) (net.Conn, error) {
	if len(sessionID) != SessionIDLen {
		return nil, fmt.Errorf("%w: session id must be %d bytes", ErrRelay, SessionIDLen)
	}
	host, port, err := ParseRelayEndpoint(rawURL)
	if err != nil {
		return nil, err
	}
	d := net.Dialer{}
	conn, err := d.DialContext(ctx, "tcp", net.JoinHostPort(host, fmt.Sprintf("%d", port)))
	if err != nil {
		return nil, err
	}
	if deadline, ok := ctx.Deadline(); ok {
		_ = conn.SetDeadline(deadline)
	}
	if _, err := conn.Write(sessionID); err != nil {
		_ = conn.Close()
		return nil, err
	}
	_ = conn.SetDeadline(time.Time{})
	return conn, nil
}

// RelayConfigured returns the trimmed connect_relay_url when present and valid,
// or "" when unset. Fail-closed on a malformed URL.
func RelayConfigured(raw string) (string, error) {
	text := strings.TrimSpace(raw)
	if text == "" {
		return "", nil
	}
	if _, _, err := ParseRelayEndpoint(text); err != nil {
		return "", err
	}
	return text, nil
}

// RelayConnHandler is invoked once a relay dial completes for a session id.
// The handler owns the connection for the session lifetime; returning closes it.
type RelayConnHandler func(ctx context.Context, conn net.Conn)

// RelaySupervisorOpts configures the outbound relay dialer supervisor.
type RelaySupervisorOpts struct {
	URL      string
	Home     string
	Handler  RelayConnHandler
	Interval time.Duration // how often to refresh wanted SIDs; default 1s
	DialWait time.Duration // per-dial timeout; default 20s
}

// RunRelaySupervisor keeps one outbound relay waiter per active rendezvous sid
// (pair window + paired devices). Matches Python _relay_supervisor: reconnect
// with backoff, skip while paused, retire SIDs that leave the wanted set.
func RunRelaySupervisor(ctx context.Context, opts RelaySupervisorOpts) error {
	url, err := RelayConfigured(opts.URL)
	if err != nil {
		return err
	}
	if url == "" {
		return nil
	}
	if opts.Handler == nil {
		return fmt.Errorf("%w: relay supervisor needs a handler", ErrRelay)
	}
	interval := opts.Interval
	if interval <= 0 {
		interval = time.Second
	}
	dialWait := opts.DialWait
	if dialWait <= 0 {
		dialWait = 20 * time.Second
	}

	type liveDial struct {
		cancel context.CancelFunc
		done   <-chan struct{}
	}
	live := map[string]liveDial{}
	defer func() {
		for _, d := range live {
			d.cancel()
		}
		for _, d := range live {
			<-d.done
		}
	}()

	tick := time.NewTicker(interval)
	defer tick.Stop()
	for {
		wanted, err := RendezvousSIDs(opts.Home)
		if err != nil {
			return err
		}
		wantKeys := map[string][]byte{}
		for _, sid := range wanted {
			if len(sid) == SessionIDLen {
				wantKeys[string(sid)] = sid
			}
		}
		for key, d := range live {
			if _, ok := wantKeys[key]; !ok {
				d.cancel()
				<-d.done
				delete(live, key)
			}
		}
		for key, sid := range wantKeys {
			if _, ok := live[key]; ok {
				continue
			}
			sidCopy := append([]byte(nil), sid...)
			dctx, cancel := context.WithCancel(ctx)
			done := make(chan struct{})
			live[key] = liveDial{cancel: cancel, done: done}
			go func() {
				defer close(done)
				runRelayDialLoop(dctx, url, opts.Home, sidCopy, dialWait, opts.Handler)
			}()
		}

		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-tick.C:
		}
	}
}

func runRelayDialLoop(ctx context.Context, url, home string, sid []byte, dialWait time.Duration, handler RelayConnHandler) {
	backoff := time.Second
	for {
		if IsPaused(home) {
			select {
			case <-ctx.Done():
				return
			case <-time.After(500 * time.Millisecond):
				continue
			}
		}
		dctx, cancel := context.WithTimeout(ctx, dialWait)
		conn, err := DialRelay(dctx, url, sid)
		cancel()
		if err != nil {
			select {
			case <-ctx.Done():
				return
			case <-time.After(backoff):
			}
			backoff = backoff + backoff/2
			if backoff > 15*time.Second {
				backoff = 15 * time.Second
			}
			continue
		}
		backoff = time.Second
		handler(ctx, conn)
		_ = conn.Close()
		select {
		case <-ctx.Done():
			return
		default:
		}
	}
}
